src/Console/CommandCollection.php

```diff
@@ -130,4 +148,60 @@ public function count()
     {
         return count($this->commands);
     }
+
+    /**
+     * Automatically discover shell commands in CakePHP, the application and all plugins.
+     *
+     * Commands will be located using filesystem conventions. Commands are
+     * discovered in the following order:
+     *
+     * - CakePHP provided commands
+     * - Application commands
+     * - Plugin commands
+     *
+     * Commands from plugins will be added based on the order plugins are loaded.
+     * Plugin shells will attempt to use a short name. If however, a plugin
+     * provides a shell that conflicts with CakePHP or the application shells,
+     * the full `plugin_name.shell` name will be used. Plugin shells are added
+     * in the order that plugins were loaded.
+     *
+     * @return array An array of command names and their classes.
+     */
+    public function autoDiscover()
+    {
+        $scanner = new CommandScanner();
+        $shells = $scanner->scanAll();
+
+        $adder = function ($out, $shells, $key) {
```

Comment:
Why not as a protected method?
I think this method could be shorter, mor readable, may be even more self-explanatory